// @hydecorp/component/lib/element.js
var _RxLitElement_firstUpdate;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { LitElement } from "lit-element";
import { Subject } from "rxjs";
export class RxLitElement extends LitElement {
    constructor() {
        super(...arguments);
        this.$connected = new Subject();
        _RxLitElement_firstUpdate.set(this, void 0);
        this.$ = {};
    }
    connectedCallback() {
        super.connectedCallback();
        this.$connected.next(true);
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        this.$connected.next(false);
    }
    firstUpdated() {
        __classPrivateFieldSet(this, _RxLitElement_firstUpdate, true, "f");
    }
    updated(changedProperties) {
        if (!__classPrivateFieldGet(this, _RxLitElement_firstUpdate, "f"))
            for (const prop of changedProperties.keys()) {
                if (prop in this.$)
                    this.$[prop].next(this[prop]);
            }
        __classPrivateFieldSet(this, _RxLitElement_firstUpdate, false, "f");
    }
    fireEvent(name, eventInitDict) {
        this.dispatchEvent(new CustomEvent(name, eventInitDict));
        this.dispatchEvent(new CustomEvent(`${this.tagName.toLowerCase()}-${name}`, eventInitDict));
    }
}
_RxLitElement_firstUpdate = new WeakMap();
export function applyMixins(derivedCtor, baseCtors) {
    baseCtors.forEach(baseCtor => {
        Object.getOwnPropertyNames(baseCtor.prototype).forEach(name => {
            derivedCtor.prototype[name] = baseCtor.prototype[name];
        });
    });
    return derivedCtor;
}
